### ISSUE TITLE
devpi test not working on Python 3.11.4

### ISSUE THREAD

**Issue by henryborchers** [devpi/devpi#978]
Ever since Python 3.11.4 came out my Linux CI machines are failing what I use the  `devpi test` command with Python 3.11.  Earlier versions of Python seem to be fine.

I get this traceback on my 3.11 machines.
```
ERROR: Exception:
Traceback (most recent call last):
  File "/tmp/devpi-test-nnanobpf/targz/HathiZip-0.1.11.dev0/.tox/py311/lib/python3.11/site-packages/pip/_internal/cli/base_command.py", line 169, in exc_logging_wrapper
    status = run_func(*args)
             ^^^^^^^^^^^^^^^
  File "/tmp/devpi-test-nnanobpf/targz/HathiZip-0.1.11.dev0/.tox/py311/lib/python3.11/site-packages/pip/_internal/cli/req_command.py", line 248, in wrapper
    return func(self, options, args)
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/tmp/devpi-test-nnanobpf/targz/HathiZip-0.1.11.dev0/.tox/py311/lib/python3.11/site-packages/pip/_internal/commands/install.py", line 375, in run
    self.trace_basic_info(finder)
  File "/tmp/devpi-test-nnanobpf/targz/HathiZip-0.1.11.dev0/.tox/py311/lib/python3.11/site-packages/pip/_internal/cli/req_command.py", line 475, in trace_basic_info
    locations = search_scope.get_formatted_locations()
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/tmp/devpi-test-nnanobpf/targz/HathiZip-0.1.11.dev0/.tox/py311/lib/python3.11/site-packages/pip/_internal/models/search_scope.py", line 82, in get_formatted_locations
    redacted_index_url = redact_auth_from_url(url)
                         ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/tmp/devpi-test-nnanobpf/targz/HathiZip-0.1.11.dev0/.tox/py311/lib/python3.11/site-packages/pip/_internal/utils/misc.py", line 527, in redact_auth_from_url
    return _transform_url(url, _redact_netloc)[0]
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/tmp/devpi-test-nnanobpf/targz/HathiZip-0.1.11.dev0/.tox/py311/lib/python3.11/site-packages/pip/_internal/utils/misc.py", line 492, in _transform_url
    purl = urllib.parse.urlsplit(url)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/urllib/parse.py", line 500, in urlsplit
    _check_bracketed_host(bracketed_host)
  File "/usr/lib/python3.11/urllib/parse.py", line 446, in _check_bracketed_host
    ip = ipaddress.ip_address(hostname) # Throws Value Error if not IPv6 or IPv4
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/ipaddress.py", line 54, in ip_address
    raise ValueError(f'{address!r} does not appear to be an IPv4 or IPv6 address')
```

Any change this is the issue?  
Changed in version 3.11.4: Leading WHATWG C0 control and space characters are stripped from the URL.
from https://docs.python.org/3/library/urllib.parse.html#urllib.parse.urlsplit

**Comment by fschulze** [comment]
That seems likely, but without the actual URL I can't investigate it. If the URL is confidential, then try to call ``urlsplit`` on it to see if you can reproduce it and then try to change the URL until you have something that causes the error but can be used in a test case.

**Comment by henryborchers** [comment]
It is a confidential url so I don't want to post it here but I'll try your suggestion when I get to work on Monday.

**Comment by henryborchers** [comment]
okay, it seems like I'm able to run `urlsplit` on what I think is being run with but it's not able to reproduce this error. Looks like I'm going to have to dig deeper.

**Comment by henryborchers** [comment]
@fschulze I was too hasty with ruling out urlsplit. It is indeed the issue. When https://github.com/python/cpython/pull/103849 was accepted.

This line was added in Python 3.11.4 https://github.com/python/cpython/pull/103849/files#diff-b3712475a413ec972134c0260c8f1eb1deefb66184f740ef00c37b4487ef873eR488

and now it break `devpi test` command when the Python interpreter is version 3.11.4.

**Comment by fschulze** [comment]
The question is why? Do you have square brackets in the URL? Are there any special characters in your URL? Do you have authentication in the URL, if so, is there a difference in your password when quoted (``assert password == urllib.parse.quote(password)``)?

**Comment by henryborchers** [comment]
I tried `assert password == urllib.parse.quote(password)` and it's false. The password includes a non-alphanumerical key

**Comment by fschulze** [comment]
Could you try:

https://m.devpi.net/fschulze/dev/devpi-client/6.0.5.dev0
https://m.devpi.net/fschulze/dev/+f/8e0/51c0ad3c869d5/devpi_client-6.0.5.dev0-py2.py3-none-any.whl

**Comment by henryborchers** [comment]
@fschulze That seemed to fix it. Nice job!

**Comment by fschulze** [comment]
I just released devpi-client 6.0.5.
